### ISSUE TITLE
Potential performance improvement in YammerPrometheusMetricsReporter

### ISSUE THREAD

**Issue by mimaison** [strimzi/metrics-reporter#26]
Currently during `collect()`, we iterate through all the metrics from the Yammer registries to apply the allow list filter.

It's possible to register listeners onto a Yammer registry. We should check if we could use this pattern and apply the filtering as metrics are added or removed instead of in `collect()`. This is similar to what's described in https://github.com/strimzi/metrics-reporter/issues/9 for `KafkaPrometheusMetricsReporter`.

**Comment by OwenCorrigan76** [comment]
@mimaison 
This is my understanding of `It's possible to register listeners onto a Yammer registry`:
The Yammer metrics library supports adding listeners that can react to changes or events related to the metrics being collected and managed by the registry. MetricsRegistry is a central place where metrics like counters, gauges etc. are registered, allowing you to manage and query the metrics efficiently. Listeners can be attached to the registry to listen for specific events such as a new metric, and the listener can handle the event. This can be done when a metric is added, removed, or updated rather than when it is collected, as the current logic is.
Currently it is handled here:
```for (MetricsRegistry registry : registries) {
            for (Map.Entry<MetricName, Metric> entry : registry.allMetrics().entrySet()) {
                System.out.println(entry.getKey() + ": key");
                System.out.println(entry.getValue() + ": value");
                MetricName metricName = entry.getKey();
                Metric metric = entry.getValue();
                // metricName is the key and metric is the value
                LOG.trace("Collecting Yammer metric {}", metricName);
                //example: kafka_server_other_type_name: prometheusMetricName
                String prometheusMetricName = metricName(metricName);
                // TODO Filtering should take labels into account
                if (!config.isAllowed(prometheusMetricName)) {
                    LOG.info("Yammer metric {} is not allowed", prometheusMetricName);
                    // skip prometheusMetricName if it is not allowed after LOG.info
                    continue;
                }
                LOG.info("Yammer metric {} is allowed", prometheusMetricName);
                Labels labels = labelsFromScope(metricName.getScope(), prometheusMetricName);
                // example {k1="v1",k2="v2"}: labels);
                LOG.info("labels {}", labels);
```
Is that a correct assumption are is there something I'm missing? Thanks

**Comment by OwenCorrigan76** [comment]
This is the relevant lines if the KafkaMetriceReporter??
https://github.com/strimzi/metrics-reporter/blob/main/src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java#L47C1-L59C6

**Comment by mimaison** [comment]
> Is that a correct assumption are is there something I'm missing? Thanks
Yes that's a good summary.

**Comment by mimaison** [comment]
> These are the relevant lines if the KafkaMetriceReporter??
> https://github.com/strimzi/metrics-reporter/blob/main/src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java#L47C1-L59C6

This ticket is about `YammerPrometheusMetricsReporter`, so I'm not sure why you're mentioning `KafkaPrometheusMetricsReporter`. Why do you think these lines are relevant to this ticket?

**Comment by OwenCorrigan76** [comment]
You referenced this abaove `This is similar to what's described in https://github.com/strimzi/metrics-reporter/issues/9 for KafkaPrometheusMetricsReporter.`

**Comment by OwenCorrigan76** [comment]
We use `import com.yammer.metrics.core.MetricsRegistryListener;` ??

**Comment by mimaison** [comment]
yes x 3

**Comment by mimaison** [comment]
Yes `MetricsRegistryListener` is the class to use but I'm really confused by the code you shared. I'm not sure I follow what you are trying to do.
